lib/SPIRV/SPIRVReader.cpp

```diff
@@ -685,49 +689,145 @@ void SPIRVToLLVM::setLLVMLoopMetadata(const LoopInstType *LM, Instruction *BI) {
+          if (SafelenIdxGroupIt == SafelenIdxGroupMap.end())
+            SafelenIdxGroupMap.emplace(std::make_pair(
+                Safelen, std::vector<MDNode *>{CurrentDepthIdxGroup}));
+          else
+            SafelenIdxGroupIt->second.push_back(CurrentDepthIdxGroup);
+
+          for (auto *GEP : ArrayGEPIt.second) {
+            StringRef IdxGroupMDName("llvm.index.group");
+            llvm::MDNode *PreviousIdxGroup = GEP->getMetadata(IdxGroupMDName);
+            if (!PreviousIdxGroup) {
+              GEP->setMetadata(IdxGroupMDName, CurrentDepthIdxGroup);
+              continue;
+            }
+
+            // If we're dealing with an embedded loop, it may be the case
+            // that the GEP instructions for some of the arrays were already
+            // marked by the algorithm when it went over the outer level loops.
+            // In order to retain the IVDep information for each "loop
+            // dimension", we will mark such GEP's into a separate joined node
+            // that will refer to the previous levels' index groups AND to the
+            // index group specific to the current loop.
+            std::vector<llvm::Metadata *> CurrentDepthOperands;
+            for (auto &Op : PreviousIdxGroup->operands())
+              CurrentDepthOperands.push_back(Op);
+            if (CurrentDepthOperands.size() == 0)
+              CurrentDepthOperands.push_back(PreviousIdxGroup);
+            CurrentDepthOperands.push_back(CurrentDepthIdxGroup);
+            auto *JointIdxGroup =
+                llvm::MDNode::get(*Context, CurrentDepthOperands);
+            GEP->setMetadata(IdxGroupMDName, JointIdxGroup);
+          }
+        }
+
+        for (auto &SflnIdxGroupIt : SafelenIdxGroupMap) {
+          auto *Name =
+              MDString::get(*Context, "llvm.loop.parallel_access_indices");
+          unsigned SflnValue = SflnIdxGroupIt.first;
+          llvm::Metadata *SafelenMDOp =
+              SflnValue ? ConstantAsMetadata::get(ConstantInt::get(
+                              Type::getInt32Ty(*Context), SflnValue))
+                        : nullptr;
+          std::vector<llvm::Metadata *> Parameters{Name};
+          for (auto *Node : SflnIdxGroupIt.second)
+            Parameters.push_back(Node);
+          if (SafelenMDOp)
+            Parameters.push_back(SafelenMDOp);
+          Metadata.push_back(llvm::MDNode::get(*Context, Parameters));
+        }
         break;
       }
+      default:
+        llvm_unreachable(
+            "Unexpected token in LoopControlExtendedConstrolsMask");
+      }
       ++NumParam;
     }
   }
   llvm::MDNode *Node = llvm::MDNode::get(*Context, Metadata);
 
   // Set the first operand to refer itself
   Node->replaceOperandWith(0, Node);
-  BI->setMetadata("llvm.loop", Node);
+  LoopObj->setLoopID(Node);
 }
 
-void SPIRVToLLVM::transLLVMLoopMetadata(const Function *F) {
+void SPIRVToLLVM::transLLVMLoopMetadata(Function *F) {
```

Comment:
That is looking fishy, why is it no longer const?

Reply:
The DominatorTree analysis does not accept a const object for reasons I was unable to determine. That said, would a `const_cast` in the `DominatorTree` object initialization make things a little better?

Reply:
I've decided to go with the `const_cast`. It used to be the choice for an analogous situation at the previous version of line  719